xds/src/main/java/io/grpc/xds/Bootstrapper.java

```diff
@@ -74,42 +79,47 @@ public static Bootstrapper getInstance() {
   /**
    * Returns configurations from bootstrap.
    */
-  public abstract BootstrapInfo readBootstrap() throws IOException;
+  public abstract BootstrapInfo readBootstrap() throws XdsInitializationException;
 
   /** Parses a raw string into {@link BootstrapInfo}. */
   @VisibleForTesting
-  @SuppressWarnings("deprecation")
-  public static BootstrapInfo parseConfig(String rawData) throws IOException {
+  @SuppressWarnings("unchecked")
+  public static BootstrapInfo parseConfig(String rawData) throws XdsInitializationException {
     XdsLogger logger = XdsLogger.withPrefix(LOG_PREFIX);
     logger.log(XdsLogLevel.INFO, "Reading bootstrap information");
-    @SuppressWarnings("unchecked")
-    Map<String, ?> rawBootstrap = (Map<String, ?>) JsonParser.parse(rawData);
+    Map<String, ?> rawBootstrap;
+    try {
+      rawBootstrap = (Map<String, ?>) JsonParser.parse(rawData);
+    } catch (IOException e) {
```

Comment:
You are not catching `ClassCastException` here e.g. when `JsonParser.parse` returns something other than a `Map` say when the input is `[3, 4, 5]`. Why not wrap that also in `XdsInitializationException`?

Reply:
That's wrong. The API caller is not expected to catch unchecked exceptions thrown by the API implementation. That breaks the purpose of abstraction. If the API implementation expects its callers to do something for the exception it throws, it should throw a checked exception. Otherwise, the burden should not be delegated to the API caller (aka, API caller is not expected to look into the API implementation).

Reply:
I am just looking at the value to the end-user and not the API abstraction or checked vs unchecked exceptions because not having the correct top level JSON structure is still an `XdsInitializationException` just like all other JSON parsing errors. It is possible to do something like:
```
Object obj = JsonParser.parse(rawData);
if (!(obj instanceof Map)) {
   throw new XdsInitializationException(...);
}
Map<String, ?> rawBootstrap = (Map<String, ?>) obj;
```

Reply:
> not having the correct top level JSON structure is still an `XdsInitializationException`

It's a bad practice to wrap runtime (unchecked) exceptions, they should be propagated straight to the program surface. The library's top-level API consumer has the error handler to do this properly. For example, `NameResolver#Listener2.onError()` is the place to gracefully propagate exceptions as error status to the channel, which eventually present to the application. So we have

https://github.com/grpc/grpc-java/blob/af6fbf6b7448768d214f8df1888eb7064bd3c650/xds/src/main/java/io/grpc/xds/XdsNameResolver.java#L119-L122

Instead of wrapping in each implementation layers, runtime exceptions should be directly forward from where it happens to where it is accepted. Again, this is because the API implementation does not expect its caller to catch it and do something, but gRPC as a library is propagating exceptions nicely to consuming applications.